packages/dynamic_layouts/lib/src/staggered_layout.dart

```diff
@@ -0,0 +1,279 @@
+  /// if not.
+  @override
+  bool reachedTargetScrollOffset(double targetOffset) {
+    for (final double scrollOffset in _scrollOffsetForMainAxis.values) {
+      if (scrollOffset < targetOffset) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  /// The infinite extent in the main axis lets
+  /// [RenderDynamicSliverGrid.performLayout()] know that the child will decide
+  /// its own size in that axis.
+  @override
+  DynamicSliverGridGeometry getGeometryForChildIndex(int index) {
+    return DynamicSliverGridGeometry(
+      scrollOffset: 0.0,
+      crossAxisOffset: 0.0,
+      mainAxisExtent: double.infinity,
+      crossAxisExtent: childCrossAxisExtent,
+    );
+  }
+
+  /// Returns an updated [DynamicSliverGridGeometry] with the actual size of
+  /// the child.
+  @override
+  DynamicSliverGridGeometry updateGeometryForChildIndex(
+      int index, Size childSize) {
+    final int crossAxisSlot = _getNextCrossAxisSlot();
+    final double currentScrollOffset = _scrollOffsetForMainAxis[crossAxisSlot]!;
+    final double childMainAxisExtent =
+        scrollDirection == Axis.vertical ? childSize.height : childSize.width;
+    _scrollOffsetForMainAxis[crossAxisSlot] =
+        childMainAxisExtent + _scrollOffsetForMainAxis[crossAxisSlot]!;
+
+    final DynamicSliverGridGeometry sliverGridGeometry =
+        DynamicSliverGridGeometry(
+      scrollOffset: currentScrollOffset +
+          _mainAxisCount[crossAxisSlot]! * mainAxisSpacing,
+      crossAxisOffset:
+          crossAxisSlot * (childCrossAxisExtent + crossAxisSpacing),
+      mainAxisExtent:
+          scrollDirection == Axis.vertical ? childSize.height : childSize.width,
+      crossAxisExtent: childCrossAxisExtent,
+    );
+    _mainAxisCount[crossAxisSlot] = _mainAxisCount[crossAxisSlot]! + 1;
+    return sliverGridGeometry;
+  }
+}
+
+/// Creates dynamic grid layouts with a fixed number of tiles in the cross axis
+/// and varying main axis size. It uses the same logic as
+/// [SliverGridDelegateWithFixedCrossAxisCount], but with a
+/// [SliverGridStaggeredTileLayout].
+///
+/// For example, if the grid is vertical, this delegate will create a layout
+/// with a fixed number of columns. If the grid is horizontal, this delegate
+/// will create a layout with a fixed number of rows.
+///
+/// This delegate creates grids with equally sized and spaced tiles.
+///
+/// See also:
+///
+///  * [DynamicSliverGridDelegateWithMaxCrossAxisExtent], which creates a
+///    dynamic layout with tiles that have a maximum cross-axis extent
+///    and varying main axis size.
+///  * [SliverGridDelegate], which creates arbitrary layouts.
+///  * [DynamicGridView], which can use this delegate to control the layout of
+///    its tiles.
+///  * [DynamicSliverGrid], which can use this delegate to control the layout
+///    of its tiles.
+///  * [RenderDynamicSliverGrid], which can use this delegate to control the
+///    layout of its tiles.
```

Comment:
All of these are stating the same thing. Maybe you could mention at which point in the layout process these each come into play?